File: bindings/rust/transcribe-cpp/examples/backend-select.rs
//! backend-select — device discovery, explicit `backend=`, graceful failure.
//!
//!     cargo run --example backend-select -- [model.gguf]
//!
//! Device discovery needs no model and always runs. With a model it shows an
//! explicit, satisfiable request (`Backend::Auto`) and, to demonstrate the
//! degradation contract, an explicit request for a backend this build lacks —
//! which must error cleanly from the request path, not crash.

#[path = "common/mod.rs"]
mod common;

use transcribe_cpp::{
    backend_available, devices, init_backends_default, Backend, Model, ModelOptions,
};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Register backend modules before anything else. A no-op in a compiled-in
    // build; in a `dynamic-backends` build it loads the per-ISA CPU / GPU
    // modules so the device list below reflects what actually registered. Must
    // happen once, before the first model load.
    init_backends_default()?;

    println!("registered compute devices:");
    for d in devices() {
        println!("  {} [{}] — {}", d.name, d.kind, d.description);
    }
    println!("\nbackend availability:");
    for b in [Backend::Cpu, Backend::Metal, Backend::Vulkan, Backend::Cuda] {
        println!("  {b:?} = {}", backend_available(b));
    }

    // The first optional backend this build does NOT have — used below to show
    // a request that cannot be satisfied. (CPU is always present, so it is not
    // a candidate.)
    let unavailable = [Backend::Cuda, Backend::Vulkan, Backend::Metal]
        .into_iter()
        .find(|&b| !backend_available(b));

    let mut args = std::env::args().skip(1);
    let Some(model_path) = common::model_path(args.next().as_deref()) else {
        eprintln!("\nskip the model half of backend-select: no model found");
        return Ok(());
    };

    // Explicit, satisfiable request: Auto resolves to the best available device.
    let model = Model::load_with(
        &model_path,
        &ModelOptions {
            backend: Backend::Auto,
            ..Default::default()
        },
    )?;
    println!(
        "\nloaded with Backend::Auto -> bound backend: {}",
        model.backend()
    );
    drop(model);

    // Graceful failure: requesting an unavailable backend must error cleanly.
    match unavailable {
        Some(b) => match Model::load_with(
            &model_path,
            &ModelOptions {
                backend: b,
                ..Default::default()
            },
        ) {
            Ok(_) => println!("requesting {b:?} unexpectedly succeeded"),
            Err(e) => println!("requesting {b:?} failed cleanly: {e}"),
        },
        None => println!("every optional backend is available on this build"),
    }
    Ok(())
}

File: bindings/typescript/examples/backend-select.mjs
// backend-select — discover devices, select exactly, show policy failure.
//
//   node examples/backend-select.mjs [model.gguf]
//
// Device discovery needs no model; model loads are skipped when none is given.

import { getAvailableBackends, backendAvailable, TranscribeModel } from "../dist/index.js";
import { model, skip } from "./_support.mjs";

const devices = getAvailableBackends();
console.log("discovered devices:");
for (const d of devices) {
  console.log(`  ${d.kind.padEnd(7)} ${d.name} — ${d.description}`);
}
console.log("\nbackend availability:");
for (const b of ["cpu", "metal", "vulkan", "cuda", "rocm"]) {
  console.log(`  ${b.padEnd(7)} ${backendAvailable(b)}`);
}

const path = model("TRANSCRIBE_SMOKE_MODEL");
if (!path) skip("\nno model — device discovery only (set TRANSCRIBE_SMOKE_MODEL to load)");

// Exact selection passes an object returned by getAvailableBackends(), not its
// display index. CPU provides a deterministic example on every platform.
const cpu = devices.find((device) => device.deviceType === "cpu");
if (!cpu) skip("no selectable CPU device");
const exact = await TranscribeModel.load(path, { device: cpu });
console.log(`\nselected exact device ${exact.device.name}, bound to: ${exact.backend}`);
exact.dispose();

// An explicit backend policy that cannot be satisfied fails cleanly.
const unavailable = ["cuda", "rocm", "vulkan", "metal"].find(
  (backend) => !backendAvailable(backend),
);
if (unavailable) {
  try {
    const unexpected = await TranscribeModel.load(path, { backend: unavailable });
    unexpected.dispose();
    console.log(`requesting ${unavailable} unexpectedly succeeded`);
  } catch (error) {
    console.log(`requesting ${unavailable} failed cleanly: ${error.message}`);
  }
} else {
  console.log("every optional backend is available on this build");
}
